matscipy/calculators/pair_potential/calculator.py

```diff
@@ -397,13 +408,98 @@ def calculate_hessian_matrix(self, atoms, H_format="dense"):
                 de_n[mask] = df[pair](abs_dr_n[mask])
                 dde_n[mask] = df2[pair](abs_dr_n[mask])
 
+        if limits != None:
+            if limits[1] < limits[0]:
+                raise ValueError(
+                    "Value error: The upper atom id cannot be smaller than the lower atom id.")
+            else:
+                mask = np.logical_and(i_n >= limits[0], i_n < limits[1])
+                i_n = i_n[mask]
+                i_n1 = i_n - i_n[0]
+                j_n = j_n[mask]
+                dr_nc = dr_nc[mask]
+                abs_dr_n = abs_dr_n[mask]
+                e_n = e_n[mask]
+                de_n = de_n[mask]
+                dde_n = dde_n[mask]
+                nat1 = limits[1] - limits[0]
+
+                first_i = [0] * (nat1 + 1)
+                j = 1
+                for k in range(1, len(i_n)):
+                    if i_n[k] != i_n[k-1]:
+                        first_i[j] = k
+                        j = j+1
+                first_i[-1] = len(i_n)
+
+                if H_format == "sparse":
+                    # Off-diagonal elements of the Hessian matrix
+                    e_nc = (dr_nc.T / abs_dr_n).T
+                    H_ncc = -(dde_n * (e_nc.reshape(-1, 3, 1)
+                                       * e_nc.reshape(-1, 1, 3)).T).T
+                    H_ncc += -(de_n / abs_dr_n * (np.eye(3, dtype=e_nc.dtype)
+                                                  - (e_nc.reshape(-1, 3, 1) * e_nc.reshape(-1, 1, 3))).T).T
+
+                    H_nat1nat = bsr_matrix(
+                        (H_ncc, j_n, first_i), shape=(3*nat1, 3*nat))
+
+                    # Stack matrices in order to obtain full shape (3*nat, 3*nat)
+                    H = vstack([bsr_matrix((limits[0]*3, 3*nat)), H_nat1nat,
+                                bsr_matrix((3*nat - limits[1]*3, 3*nat))])
+
+                    # Diagonal elements of the Hessian matrix
+                    Hdiag_icc = np.empty((nat1, 3, 3))
+                    for x in range(3):
+                        for y in range(3):
+                            Hdiag_icc[:, x, y] = - \
+                                np.bincount(i_n1, weights=H_ncc[:, x, y])
+
+                    Hdiag_nat1nat = bsr_matrix((Hdiag_icc, np.arange(limits[0], limits[1]),
+                                                np.arange(nat1+1)), shape=(3*nat1, 3*nat))
+
+                    # Compute full Hessian matrix
+                    H += vstack([bsr_matrix((limits[0]*3, 3*nat)), Hdiag_nat1nat,
+                                 bsr_matrix((3*nat - limits[1]*3, 3*nat))])
+
+                    return H
+
+                elif H_format == "dense":
+                    # Off-diagonal elements of the Hessian matrix
```

Comment:
The same is actually true for the "sparse" vs "dense" separation. Do the codes need to be entirely separate?